lib/common/strings.c

```diff
@@ -782,3 +783,66 @@ pcmk_numeric_strcasecmp(const char *s1, const char *s2)
     }
     return 0;
 }
+
+int
+pcmk__strcmp(const char *s1, const char *s2, uint32_t flags)
+{
+    /* If this flag is set, the second string is a regex. */
+    if (is_set(flags, pcmk__str_regex)) {
+        regex_t *r_patt = calloc(1, sizeof(regex_t));
+        int reg_flags = REG_EXTENDED | (is_set(flags, pcmk__str_casei) ? REG_ICASE : 0);
+        int regcomp_rc = 0;
+        int rc = 0;
+        int nregs = 0;
+        regmatch_t *pmatch = NULL;
+
+        regcomp_rc = regcomp(r_patt, s2, reg_flags);
+        if (regcomp_rc != 0) {
+            rc = 1;
+            crm_err("Bad regex '%s' for update: %s", s2, strerror(regcomp_rc));
+        } else {
+            rc = regexec(r_patt, s1, nregs, pmatch, 0);
+        }
+
+        free(pmatch);
+        regfree(r_patt);
+        free(r_patt);
+
+        /* regexec returns 0 for a match, and some constant for no match.  We can just
+         * repurpose that non-zero return code here.  It doesn't line up exactly with
+         * how strcmp works, but the caller will know they used a regex and can expect
+         * this return code.
+         */
```

Comment:
I'm thinking we should return either -1 or 1 for no match, so the caller doesn't have to include regex.h to compare against REG_NOMATCH. It doesn't really matter which we pick as long as we document it. I expect callers will just use ==/!= 0 anyway, but it's nice to have well-defined behavior.

Reply:
I'm already returning 1 for when the user passes a bad regex, so I might as well just use 1 to indicate any error condition (as long as we don't care about the callers being able to differentiate).  My expectation was that peopel would use == 0 or != 0 as well, in which case it didn't much matter what the return value was.  But we might as well define what's going to happen.